### DISCUSSION TITLE
Error: JSON: Syntax error, malformed JSON.

### DISCUSSION THREAD

**Discussion by fritzisa** [code4recovery/12-step-meeting-list#782] [Q&A]
**Dear Code4Recovery friends,**

I have (2) feeds that pass the validator but give me an error message when Refreshing or Adding Data Source.

District 42 (VALID)
https://padistrict42.org/wp-admin/admin-ajax.php?action=meetings&key=854aec833cc75bb085e64a72b899a2bb

District 35 (VALID)
https://www.d35pa-aa.org/wp-admin/admin-ajax.php?action=meetings&key=c092a973561757a0f800a6bc5822dd2

Any assistance would be greatly appreciated.

Thank you,

**@scott_f**
Area 59 Web Servant
https://area59.aa.org

<img width="1254" alt="Screen Shot 2022-04-22 at 4 05 25 PM" src="https://user-images.githubusercontent.com/95769327/164786892-45d02550-894a-4d6e-9d70-a23eba512bd2.png">
a

**Comment by joshreisner** [comment]
i think my best answer to this is still https://github.com/code4recovery/12-step-meeting-list/discussions/724#discussioncomment-2348163

**Reply by kiyote33** [reply]
For what its' worth, I've added both District 42 (35 records) and District 35 (115 records) to my sandbox without any errors. Is this a staging site or live production. It seems like it may be a local site issue.

**Reply by fritzisa** [reply]
Thank you for checking.
Previously Josh had replied: 

> Do you have command line access on your website (the one you're importing from)?
> 
> If so, try running curl "https://www.d35pa-aa.org/wp-admin/admin-ajax.php?action=meetings&key=c092a973561757a0f800a6bc5822dd2a" and see if you see json.
> 
> The only thing I can think of is if there's some connection problem between your two websites.
> 
> It's extra strange since you can import other feeds without incident.

I guess I will need to pass that along to the web servants of the two Districts we're having the problem with. (?)

I have supplied all our District and Intergroup web servants with the attached instruction. Maybe there's a setting in there that they've changed? Both of these Districts' feeds had worked for months before we started having problems.

Any additional input would be appreciated
[Meeting-Feed-Creation-MARCH-2022.pdf](https://github.com/code4recovery/12-step-meeting-list/files/8544391/Meeting-Feed-Creation-MARCH-2022.pdf)
.

**Reply by ZamulaK** [reply]
Hi @kiyote33 and @joshreisner - this is Dan B from Area 59. I'm the current Area Treasurer but do programming for a living, so I've starting assisting with some of the technical issues around the plugin and JSON feeds.

For the issue Scott identified, we've come up with a work-around as we continue to investigate the root cause. The summary so far:
- Our workaround is to use a [Google Apps Script](https://script.google.com/a/area59aa.org/macros/s/AKfycbySkcu1R3oO2P-R76bPzPx2ZnMtT0RY-02OmEKZUA/exec?id=35,42) that pulls directly from the District 42 and 35 feeds, and then outputs it again as a JSON feed.
  - This points to a connection issue from our site to the source feeds, rather than it being a data issue.
  - To confirm it's not a data issue, we saved the raw JSON responses to a file after making REST requests to the problematic source feeds, and then uploaded those to GitHub. We could make the GitHub files a feed source for the plugin without any errors.
- In reviewing the source code for the plugin, it looks like the syntax error is being thrown from the call to `json_decode()`:
  -  It appears this block evaluates to false: 
`if (is_array($response) && !empty($response['body']) && ($body = json_decode($response['body'], true))) {`
  - And then the final `else` condition has a `switch()` statement with the error check:
`case JSON_ERROR_SYNTAX:
   tsml_alert(__('JSON: Syntax error, malformed JSON.', '12-step-meeting-list'), 'error');`
  - Would turning on server-level debugging provide any additional information about the actual error?
  - Is there any way to get the raw response being returned from the call to `wp_remote_get()`?
 
I'm not familiar with WordPress or PHP development (yet), but probably will install PHP to gain from experience, and possibly contribute in the future :) 

Any insights on how to proceed from here would be greatly appreciated!

**Reply by joshreisner** [reply]
hi @ZamulaK ! yes i think we've confirmed that the problem is not with the feeds themselves, since we can import them to our website fine.  in @fritzisa's previous post i asked:

> Do you have command line access on your website (the one you're importing from)? If so, try running `curl "https://www.d35pa-aa.org/wp-admin/admin-ajax.php?action=meetings&key=c092a973561757a0f800a6bc5822dd2a"` and see if you see json.

if you had that this would tell us something! if not you could try uploading a file to your theme with some code to execute the `wp_remote_get()` line, i could give you some advice on how to do that

**Reply by ZamulaK** [reply]
Hi @joshreisner, thanks for the reminder about the curl command! I had forgotten we can use that to directly download the file. Although I don't have command line access to our website, the tech team is probably going to meet this week, so hopefully we can get the results.

I'll let you know if we need an alternate route, and thanks again for the help!

**Reply by tech2serve** [reply]
I'll chime in with a generic troubleshooting tip that I've seen resolve oddball issues like this in the past. We've seen too many cases where other plugins interfere with TSML, and so you might try disabling all other plugins (perhaps clone to a test site?) and then try the imports again. 

Admittedly though, if only some feeds are doing this and not all, it would be weird.

**Reply by ZamulaK** [reply]
So it looks like we currently don't have terminal access; our site is managed WordPress through GoDaddy and SSH access shows as disabled. 

Unfortunately none of use are really strong on the backend part:
- I'm the most "developer-y" but don't have PHP/WP experience.
- Our Web Servant is more of a designer then programmer. 
- If someone could walk us through the easiest way to get a PHP file to output the results of `wp_remote_get()` I'd be grateful!
  - The hardest part will probably be the steps needed to upload a PHP file into the theme directory that executes the code. After a few Google searches, it's unclear what's the best way.
  - The actual PHP code to do the `wp_remote_get()` seems relatively straightforward though.

**Reply by joshreisner** [reply]
ok here's how i would do it - there are likely to be several ways to do this though:

1. create a new "page" in wordpress called "test" -- it doesn't need any content, make sure the the permalink is `test` (or if that's already taken, remember what the permalink was)
2. navigate to this page, you should see a more-or-less empty page, except navigation and site branding
3. create a new text file in your active theme folder (via FTP or File Manager). call it `page-test.php` (or whatever your page slug was)
4. hit refresh in your browser, now the page should be purely blank white
5. now you should be able to drop some code in the page and have it running inside wordpress
6. try adding following code, and then hit refresh:

```
<pre>
<?php
print_r(wp_remote_get('https://www.d35pa-aa.org/wp-admin/admin-ajax.php?action=meetings&key=c092a973561757a0f800a6bc5822dd2a'));
```

**Reply by joshreisner** [reply]
i followed these steps on our website: https://code4recovery.org/test

you can ignore most of it but the important thing is what's in the line that starts with `[body] => ` -- it shows valid JSON

**Reply by ZamulaK** [reply]
Hi @joshreisner,

So I managed to get a couple of test pages up, if you'd like to take a look:

- [https://area59aa.org/testfeed1/](https://area59aa.org/testfeed1/)
  This is your original code.
- [https://area59aa.org/testfeed2/](https://area59aa.org/testfeed2/)
  This decodes the JSON, with the same line as in the plugin:
  ```
  <pre>
  <?php
   print_r(json_decode(wp_remote_get('https://padistrict42.org/wp-admin/admin-ajax.phpaction=meetings&key=854aec833cc75bb085e64a72b899a2bb')['body'], true));
  ````
The really strange thing is the JSON looks valid - not at all what was expected! Maybe you'll see something of note.

**Reply by joshreisner** [reply]
yeah, it looks totally valid. what a mystery. you're still having problems importing? what if you remove and re-add pa district 42?
